singer/src/main/java/com/pinterest/singer/reader/mapped/MappedFileTBinaryProtocol.java

```diff
@@ -0,0 +1,407 @@
+   * @throws IOException
+   */
+  protected ByteBuffer mapFile(File file) throws FileNotFoundException, IOException {
+    RandomAccessFile raf = new RandomAccessFile(file, "r");
+    MappedByteBuffer map = raf.getChannel().map(MapMode.READ_ONLY, 0, file.length());
+    raf.close();
+    return map;
+  }
+
+  /**
+   * Reading methods.
+   */
+
+  public TMessage readMessageBegin() throws TException {
+    int size = readI32();
+    if (size < 0) {
+      int version = size & VERSION_MASK;
+      if (version != VERSION_1) {
+        throw new TProtocolException(TProtocolException.BAD_VERSION,
+            "Bad version in readMessageBegin");
+      }
+      return new TMessage(readString(), (byte) (size & 0x000000ff), readI32());
+    } else {
+      if (strictRead) {
+        throw new TProtocolException(TProtocolException.BAD_VERSION,
+            "Missing version in readMessageBegin");
+      }
+      return new TMessage(readStringBody(size), readByte(), readI32());
+    }
+  }
+
+  public void readMessageEnd() {
+  }
+
+  public TStruct readStructBegin() {
+    try {
+      // since data is framed, each message read is preceded by a 4 byte size
+      readI32();
+    } catch (TException e) {
+      logger.log(Level.SEVERE, "Failed to read frame", e);
+    }
+    return ANONYMOUS_STRUCT;
+  }
+
+  public void readStructEnd() {
+  }
+
+  public TField readFieldBegin() throws TException {
+    byte type = readByte();
+    short id = type == TType.STOP ? 0 : readI16();
+    return new TField("", type, id);
+  }
+
+  public void readFieldEnd() {
+  }
+
+  public TMap readMapBegin() throws TException {
+    TMap map = new TMap(readByte(), readByte(), readI32());
+    checkContainerReadLength(map.size);
+    return map;
+  }
+
+  public void readMapEnd() {
+  }
+
+  public TList readListBegin() throws TException {
+    TList list = new TList(readByte(), readI32());
+    checkContainerReadLength(list.size);
+    return list;
+  }
+
+  public void readListEnd() {
+  }
+
+  public TSet readSetBegin() throws TException {
+    TSet set = new TSet(readByte(), readI32());
+    checkContainerReadLength(set.size);
+    return set;
+  }
+
+  public void readSetEnd() {
+  }
+
+  public boolean readBool() throws TException {
+    return (readByte() == 1);
+  }
+
+  public byte readByte() throws TException {
+    checkAvailable(1);
+    return buffer.get();
+  }
+
+  /**
+   * Checks if the specified number of bytes are available to be read from the
+   * byte buffer. If this is not true the method checks if the file has more data
+   * to be read and will trigger a remap of the file.
+   *
+   * @param size
+   * @throws TException
+   */
+  protected void checkAvailable(int size) throws TException {
+    if (buffer.remaining() < size) {
+      if (file.length() > buffer.capacity()) {
+        remap();
+      } else {
+        throw new TException(
+            "No more data available: " + file.length() + " vs." + buffer.capacity());
```

Comment:
Might help if we identify which is which in the message (length vs. capacity).

Reply:
Fixed